### DISCUSSION TITLE
Some IMF verification problems when transwrapping video MXF files

### DISCUSSION THREAD

**Discussion by JoakimZiegler** [bbc/bmx#85] [General]
I'm doing some testing on creating HDR10 IMF packages from Davinci Resolve. This is a bit tricky, because Resolve doesn't allow setting of all the relevant metadata that needs to go in the CPL. Notably, I'm creating a video track which is P3D65 limited but in a rec.2020 container, which Resolve writes out with metadata indicating just rec.2020. Additionally, I need to add HDR metadata (MaxFALL and MaxCLL) to the CPL.

To this end, I find myself doing some manual surgery on the files in the IMF. Adding the metadata to the CPL is simple, I just need to update the other XML files in the IMF to reflect the new hash values, etc., but when I try to update the Mastering Monitor Color Volume settings, this clashes with the metadata in the video MXF file, and Photon validation fails.

This all lead me to finding bmx, which I have to say is pretty great, and promises to do exactly what I need, in that I can transwrap the video MXF file while modifying its metadata, so that it'll correspond to the metadata I'm setting in the CPL. In my test, I used it like this:

`bmxtranswrap  -t imf --display-primaries 34000,16000,13250,34500,7500,3000 --display-white-point 15635,16450 --display-max-luma 10000000 --display-min-luma 1 -o {Type}_{fp_uuid}.mxf TEST_IMF/VIDEO_13179de4-4e86-4928-a114-ab83a4568d4f.mxf`

And then I swapped out the original MXF file for the new one, calculated its hash, added that to the relevant XML files, updated everything so it matches, and, lo and behold, Photon validation kind of works for this, in that the Mastering Monitor Color Volume settings now match and verify correctly, but it fails on some new metadata I've not seen it complain about before. This is what Photon says:

```25 Jul 2024 09:10:24,421 [INFO ] [IMPAnalyzer]: ==========================================================================
25 Jul 2024 09:10:24,421 [INFO ] [IMPAnalyzer]: Virtual Track Conformance
25 Jul 2024 09:10:24,421 [INFO ] [IMPAnalyzer]: ==========================================================================
25 Jul 2024 09:10:24,426 [INFO ] [IMPAnalyzer]: CPL_39f3fa6d-c38f-49ac-bc7b-0c359c5254e2.xml Virtual Track Conformance has 1 errors and 0 warnings
25 Jul 2024 09:10:24,426 [ERROR] [IMPAnalyzer]: 		ERROR-EssenceDescriptor with Id ab0a1fe7-752e-494f-bcd9-b4b5c92cf472 in the CPL's EssenceDescriptorList doesn't match any EssenceDescriptors within the IMFTrackFile resource that references it,

EssenceDescriptor in CPL EssenceDescriptorList with mismatching fields is as follows
<RGBADescriptor xmlns="http://www.smpte-ra.org/reg/395/2014/13/1/aaf">
<LinkedTrackID xmlns="http://www.smpte-ra.org/reg/335/2012">2</LinkedTrackID>
<ActiveFormatDescriptor xmlns="http://www.smpte-ra.org/reg/335/2012">66</ActiveFormatDescriptor>
<InstanceID xmlns="http://www.smpte-ra.org/reg/335/2012">urn:uuid:d7c68e12-f4fd-447a-930d-9b38add4201c</InstanceID>
<SubDescriptors xmlns="http://www.smpte-ra.org/reg/335/2012">
<JPEG2000SubDescriptor xmlns="http://www.smpte-ra.org/reg/395/2014/13/1/aaf">
<InstanceID xmlns="http://www.smpte-ra.org/reg/335/2012">urn:uuid:4cc69189-c87f-4675-8edc-4a14b5e1d00f</InstanceID>
</JPEG2000SubDescriptor>
</SubDescriptors>
</RGBADescriptor>
,

EssenceDescriptor found in the TrackFile resource with mismatching fields is as follows
<RGBADescriptor xmlns="http://www.smpte-ra.org/reg/395/2014/13/1/aaf">
<LinkedTrackID xmlns="http://www.smpte-ra.org/reg/335/2012">1001</LinkedTrackID>
<ActiveFormatDescriptor xmlns="http://www.smpte-ra.org/reg/335/2012">68</ActiveFormatDescriptor>
<InstanceID xmlns="http://www.smpte-ra.org/reg/335/2012">urn:uuid:218518c2-e37b-4947-b9c7-74aede25420f</InstanceID>
<SubDescriptors xmlns="http://www.smpte-ra.org/reg/335/2012">
<JPEG2000SubDescriptor xmlns="http://www.smpte-ra.org/reg/395/2014/13/1/aaf">
<InstanceID xmlns="http://www.smpte-ra.org/reg/335/2012">urn:uuid:60dbd832-e1e1-4f7f-a156-a73da9a9daf7</InstanceID>
</JPEG2000SubDescriptor>
</SubDescriptors>
</RGBADescriptor>


[Photon version: 4.8.0]
```

Now, I have no idea where these UUIDs are coming from, nor why the ActiveFormatDescriptor and LinkedTrackIDs are different. So, my question is, how can I get the correct UUIDs to use in my CPL EssenceDescriptorList so that this will check out, and how can I make bmx transwrap copy the ActiveFormatDescriptor and LinkedTrackID correctly, or, if not, can I just set the ActiveFormatDescriptor, for instance, in the CPL file, to what it says it is in the MXF file, assuming this is metadata that's embedded in the MXF file somehow?

Sorry for the long data dump, and thank you again for bmx, it looks like a great set of tools that I'm sure I'm going to get a lot of use out of.

**Comment by philipnbbc** [comment]
The essence descriptor in the CPL is an XML representation of the essence descriptor in the MXF file. If you re-wrap the file then the XML also needs to be updated because the re-wrap is not one-to-one.

bmx uses it's own TrackID numbering scheme, which is why the LinkedTrackID changed to 1001. I'm assuming the AspectRatio is set to 16:9 and that will explain why bmx fixes the ActiveFormatDescriptor value to be 68. The InstanceID properties are low-level identifiers for the metadata sets in the MXF file that change everytime a new file is written. You should be able to manually change the values to match what's in the new MXF file.

There are a couple open source IMF tools listed [here](https://www.imfug.com/open-source/) which should help you (Photon is listed as well). There is also the [SMPTE MXF Dump online tool](https://registry.smpte-ra.org/apps/regxmldump/view/draft/) that you can use to experiment with to test what you need to get from the other tools.

**Reply by JoakimZiegler** [reply]
Thank you, this is a very helpful starting point to figure out what's going on, and it's good to know that it should work to just update the CPL file to match the new values. The aspect ratio is indeed 16:9, which makes me wonder why the value wasn't set properly in the first place.

What's ActiveFormatDescriptor 66 equivalent to? And, where can I find a list of these? I will occasionally need to create IMFs with letterboxed or pillarboxed content, so it would be good to know. Is it possible to set the ActiveFormatDescriptor with bmx while transwrapping?

**Reply by philipnbbc** [reply]
The ActiveFormatDescriptor is defined in the MXF spec, SMPTE ST 377-1 section G.2.5, which references SMPTE 2016-1 section 5. There is also a description of ActiveFormatDescriptor available [here](https://qc.ebu.io/items/0001W/) if you don't have the SMPTE specs. and the [DVB spec](https://www.etsi.org/deliver/etsi_ts/101100_101199/101154/01.08.01_60/ts_101154v010801p.pdf) has some information about AFD values in section B.3

The bits for ActiveFormatDescriptor are defined as follows:
* bit 7 - ignore
* bit 3 to bit 6 - the AFD value defined in SMPTE 2016-1 section 5
* bit 2 - aspect ratio, 1 = 16:9, 0 = 4:3
* bit 0 to bit 1 = should be ignored by decoders

Both 66 and 68 have an AFD value 8 (full frame image). So a full frame image for 16:9 = 8<<3|1<<2 = 68.

You can use the `--afd` option to set the AFD value. bmx will then use that and the aspect ratio to construct the ActiveFormatDescriptor value.

**Reply by JoakimZiegler** [reply]
Excellent, thank you for your help.

**Reply by JoakimZiegler** [reply]
If anyone runs into this discussion later, [Wikipedia has a full list of AFD codes](https://en.wikipedia.org/wiki/Active_Format_Description). Looks like for most purposes, 68 is fine if you're delivering 16:9, there's no allowance for letterbox ratios wider than 16:9, which was my main concern.
